File: backend/app/schemas/booking.py
from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.booking import BookingStatus
from app.models.vehicle import FuelType, VehicleType


class BookingCreate(BaseModel):
    """The mechanic is identified by their public profile id, which is what
    discovery (FR-02) returns. The customer is taken from the bearer token."""

    mechanic_profile_id: UUID
    # Chosen in the booking flow; no saved vehicle required.
    vehicle_type: VehicleType
    problem_description: str = Field(..., min_length=5, max_length=2000)
    service_latitude: float = Field(..., ge=-90.0, le=90.0)
    service_longitude: float = Field(..., ge=-180.0, le=180.0)
    service_address: Optional[str] = Field(None, max_length=500)


class BookingVehicle(BaseModel):
    id: UUID
    vehicle_type: VehicleType
    brand: str
    model: str
    fuel_type: FuelType
    registration_number: Optional[str] = None
    nickname: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class BookingParty(BaseModel):
    """Minimal counterpart details. Contact data is intentionally limited."""

    name: str
    phone_number: Optional[str] = None


class BookingResponse(BaseModel):
    id: UUID
    status: BookingStatus
    problem_description: str
    service_latitude: float
    service_longitude: float
    service_address: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    vehicle_type: VehicleType
    # Present only when the booking was linked to a saved vehicle.
    vehicle: Optional[BookingVehicle] = None
    customer: BookingParty
    mechanic: BookingParty

    model_config = ConfigDict(from_attributes=True)


class BookingListResponse(BaseModel):
    bookings: List[BookingResponse]

File: frontend/src/lib/booking.ts
/** Booking endpoints (RAD FR-03 / FR-05). Typed wrappers over `api.ts`. */
import { apiFetchData } from './api';
import type { VehicleType } from './mechanic';

export const BOOKING_STATUSES = [
  'PENDING', 'ACCEPTED', 'IN_PROGRESS', 'COMPLETED', 'CANCELLED', 'REJECTED',
] as const;
export type BookingStatus = (typeof BOOKING_STATUSES)[number];

export const STATUS_LABELS: Record<BookingStatus, string> = {
  PENDING: 'Awaiting mechanic',
  ACCEPTED: 'Accepted',
  IN_PROGRESS: 'Service in progress',
  COMPLETED: 'Completed',
  CANCELLED: 'Cancelled',
  REJECTED: 'Declined',
};

export const STATUS_TONES: Record<BookingStatus, string> = {
  PENDING: 'bg-amber-500/10 border-amber-500/30 text-amber-300',
  ACCEPTED: 'bg-emerald-500/10 border-emerald-500/30 text-emerald-300',
  IN_PROGRESS: 'bg-sky-500/10 border-sky-500/30 text-sky-300',
  COMPLETED: 'bg-white/5 border-white/15 text-zinc-300',
  CANCELLED: 'bg-red-500/10 border-red-500/30 text-red-300',
  REJECTED: 'bg-red-500/10 border-red-500/30 text-red-300',
};

export interface BookingVehicle {
  id: string;
  vehicle_type: VehicleType;
  brand: string;
  model: string;
  registration_number?: string | null;
  nickname?: string | null;
}

export interface BookingParty {
  name: string;
  phone_number?: string | null;
}

export interface Booking {
  id: string;
  status: BookingStatus;
  vehicle_type: VehicleType;
  problem_description: string;
  service_latitude: number;
  service_longitude: number;
  service_address?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  /** Present only when the booking was linked to a saved vehicle. */
  vehicle?: BookingVehicle | null;
  customer: BookingParty;
  mechanic: BookingParty;
}

export interface CreateBookingInput {
  mechanic_profile_id: string;
  /** Chosen during booking. No saved vehicle is required. */
  vehicle_type: VehicleType;
  problem_description: string;
  service_latitude: number;
  service_longitude: number;
  service_address?: string | null;
}

export function createBooking(input: CreateBookingInput): Promise<Booking> {
  return apiFetchData<Booking>('/bookings/', { method: 'POST', body: input });
}

/** Customers get the bookings they raised; mechanics get the ones assigned to them. */
export async function listBookings(statuses?: BookingStatus[]): Promise<Booking[]> {
  const query = statuses?.length
    ? `?${statuses.map((s) => `status=${s}`).join('&')}`
    : '';
  const result = await apiFetchData<{ bookings: Booking[] }>(`/bookings/${query}`);
  return result.bookings;
}

export function getBooking(id: string): Promise<Booking> {
  return apiFetchData<Booking>(`/bookings/${id}`);
}

/** The server owns the state machine; these are intents, not status writes. */
type BookingAction = 'accept' | 'reject' | 'start' | 'complete' | 'cancel';

export function actOnBooking(id: string, action: BookingAction): Promise<Booking> {
  return apiFetchData<Booking>(`/bookings/${id}/${action}`, { method: 'POST' });
}
